### PR TITLE
add gyro bmi160 and fix gyro i2c address

### PR DESCRIPTION
**Pull request by AutoPlantBali**
esp32 bmi160 on spi
esp32_bmi160_spi

esp32 bmi160 on i2c
esp32_bmi160_i2c

rp2040 bmi160 on spi
rp2040_bmi160_spi

fix i2c address detect 0x76 and 0x77
https://github.com/rtlopez/esp-fc/issues/81

### CODE REVIEW

File: lib/Espfc/src/Device/GyroBMI160.h [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+      if(!testConnection()) return 0;
+
+      // reset device
+      _bus->writeByte(_addr, BMI160_RA_CMD, BMI160_CMD_SOFT_RESET);
+      delay(1);
+
+      /* Issue a dummy-read to force the device into SPI comms mode */
+      uint8_t dummy = 0;
+      _bus->readByte(_addr, 0x7F, &dummy);
```
Comment (rtlopez) [inline lib/Espfc/src/Device/GyroBMI160.h:67]: avoid using magic numbers in code

File: lib/Espfc/src/Device/GyroBMI160.h [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      // // Set up full scale Accel range. +-16G
+      _bus->writeByte(_addr, BMI160_RA_ACCEL_RANGE, 0x0C);
+      delay(1);
+
+      // // Set up full scale Gyro range. +-2000dps
+      _bus->writeByte(_addr, BMI160_RA_GYRO_RANGE, 0x00);
+      delay(1);
+
+      // Set Accel ODR to 400hz, BWP mode to Oversample 4, LPF of ~40.5hz
+      _bus->writeByte(_addr, BMI160_RA_ACCEL_CONF, 0x0A);
```
Comment (rtlopez) [inline lib/Espfc/src/Device/GyroBMI160.h:92]: Does this comment match real setting? 400 Hz ODR and 40Hz LPF is actually useless, either for accel and gyro.

File: lib/Espfc/src/Device/GyroBMI160.h [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      return 1;
+    }
+
+    void setDLPFMode(uint8_t mode) override
+    {
+    }
+
+    int getRate() const override
+    {
+      return 8000;
```
Comment (rtlopez) [inline lib/Espfc/src/Device/GyroBMI160.h:139]: I haven't studied the BMI160 documentation deeply yet, but it seems that its native operating frequency is 3200Hz. It is good for ODR to be equal to gyro reading rate. So return 3200 here, and set correct ODR in `setRate()` method below. You can find proper formula in datasheet. Accel ODR is not as much important, can be fixed on frequency close to 500Hz.

**Review by rtlopez — COMMENTED** [review]
Hi @AutoPlantBali Thanks for your contribution. I'd like to ask, that did you build any craft with this gyro? can you confirm that this code works "in the air"? Can you show any blacbox recording? 

Fix for alternative i2c address looks fine. 

Cheers

**Comment by AutoPlantBali** [comment]
> Hi @AutoPlantBali Thanks for your contribution. I'd like to ask, that did you build any craft with this gyro? can you confirm that this code works "in the air"? Can you show any blacbox recording?
> 
> Fix for alternative i2c address looks fine.
> 
> Cheers

I made a board that supports mpu6500, mpu9250, bmi160, lsm6ds and bmp280. Next month I will order the PCB and start assembling it. I will show the results of testing using this sensor
esp_fc_top

File: lib/Espfc/src/Device/GyroBMI160.h (hunk 1 above)
Comment (AutoPlantBali) [reply lib/Espfc/src/Device/GyroBMI160.h:67]: I will add it to the define list

File: lib/Espfc/src/Device/GyroBMI160.h (hunk 2 above)
Comment (AutoPlantBali) [reply lib/Espfc/src/Device/GyroBMI160.h:92]: /** Get accelerometer output data rate.
 * The acc_odr parameter allows setting the output data rate of the accelerometer
 * as described in the table below.
 *
 * <pre>
 *  5 =  25/2Hz
 *  6 =    25Hz
 *  7 =    50Hz
 *  8 =   100Hz
 *  9 =   200Hz
 * 10 =   400Hz
 * 11 =   800Hz
 * 12 =  1600Hz
 * 13 =  3200Hz
 * </pre>

File: lib/Espfc/src/Device/GyroBMI160.h (hunk 3 above)
Comment (AutoPlantBali) [reply lib/Espfc/src/Device/GyroBMI160.h:139]: I will fix it by taking the response from the sensor

**Comment by rtlopez** [comment]
Hi @AutoPlantBali, I merged your changes to develop branch, as I want to do some tests first. Next time It would be better if you put different features and fixes in separate PRs. Thanks again.

**Comment by rtlopez** [comment]
Hi @AutoPlantBali I've connected brand new BMI160 gyro, and calibration not passed. It seems that gyro has stored inaccurate offset values in NVM, this caused that gyro bias for x and y axis was around 10 deg/s. If I turn off applying offset, then bias is about 1 deg/s. Can you test these changes from develop branch with your device?
https://github.com/rtlopez/esp-fc/commit/19ed00762bfc078fb22ebd4b02da8aeaf1280efb
Thanks

**Pull request merged**